File: código-fonte/Python/paralelo/memmap/correlacao.py
from multiprocessing import Process
from memoria_compartilhada import *
import numpy as np

def __aplicar_correlacao_parcial(altura_imagem : int, largura_imagem : int, inicio : int, fim : int, mascara):
    imagem_entrada = get_memoria_compartilhada(MemoriaCompartilhada.IMAGEM_ENTRADA, altura_imagem, largura_imagem)
    imagem_saida = get_memoria_compartilhada(MemoriaCompartilhada.IMAGEM_SAIDA, altura_imagem, largura_imagem)
    
    altura_mascara, largura_mascara = mascara.shape
    
    centro_largura = largura_mascara // 2
    centro_altura = altura_mascara // 2
    
    altura_mascara_par = (altura_mascara % 2 == 0)
    largura_mascara_par = (largura_mascara % 2 == 0)
    
    for i in range(inicio, fim):
        for j in range(largura_imagem):
            regiao = imagem_entrada[max(0, i-centro_altura):min(altura_imagem, i+centro_altura if altura_mascara_par else i+centro_altura+1),
                        max(0, j-centro_largura):min(largura_imagem, j+centro_largura if largura_mascara_par else j+centro_largura+1)]
            
            imagem_saida[i, j] = np.sum(regiao * mascara[max(0, centro_altura - i):min(altura_mascara, altura_imagem - (i - centro_altura)),
                                        max(0, centro_largura - j):min(largura_mascara, largura_imagem - (j - centro_largura))]).astype(np.uint8)    

def aplicar_correlacao(imagem_entrada : np.ndarray, mascara : np.ndarray, num_processo : int) -> np.ndarray:
    altura_imagem, largura_imagem = imagem_entrada.shape
    
    imagem_saida = criar_imagem_saida(altura_imagem, largura_imagem)

    linhas_por_processo : int = int(altura_imagem / num_processo)
    
    processos = [Process(target=__aplicar_correlacao_parcial,
                         args=(altura_imagem,
                               largura_imagem,
                               p * linhas_por_processo,
                               altura_imagem if (p == num_processo - 1) else (p + 1) * linhas_por_processo,
                               mascara)) for p in range(num_processo)]
    
    for p in processos:
        p.start()
        
    for p in processos:
        p.join()
        
    return imagem_saida

def gerar_mascara_media(dimensao : int):
    mascara = np.zeros((dimensao, dimensao), dtype=np.float32)
    
    media : float = 1 / float(dimensao * dimensao)
    
    mascara[:,:] = media
    
    return mascara

File: código-fonte/C++/paralelo/correlacao.cpp
#include <opencv2/opencv.hpp>
#include <thread>
#include "correlacao.hpp"

using namespace std;
using namespace cv;

void correlacao_parcial(const Mat &imagem_entrada, const Mat &mascara, Mat &imagem_saida, 
                        const int inicio, const int fim) {
    const int altura_imagem = imagem_entrada.rows;
    const int largura_imagem = imagem_entrada.cols;

    const int altura_mascara = mascara.rows;
    const int largura_mascara = mascara.cols;

    const int centro_altura = altura_mascara / 2;
    const int centro_largura = largura_mascara / 2;

    for (int i = inicio; i < fim; i++) {
        for (int j = 0; j < largura_imagem; j++) {
            float soma = 0;

            for (int m = 0; m < altura_mascara; m++) {
                for (int n = 0; n < largura_mascara; n++) {
                    int y = i - centro_altura + m;
                    int x = j - centro_largura + n;

                    if ((0 <= y && y < altura_imagem) && (0 <= x && x < largura_imagem)) {
                        soma += imagem_entrada.at<uchar>(y, x) * mascara.at<float>(m, n);
                    }
                }
            }
            imagem_saida.at<uchar>(i, j) = (uchar) soma;
        }
    }
}

Mat aplicar_correlacao(const Mat &imagem_entrada, const Mat &mascara, const int num_threads) {
    const int altura_imagem = imagem_entrada.rows;
    const int largura_imagem = imagem_entrada.cols;

    const int linhas_por_thread = altura_imagem / num_threads;
    thread threads[num_threads];

    Mat imagem_saida = Mat::zeros(altura_imagem, largura_imagem, CV_8UC1); // 8 bits uchar 1 banda

     // Calcular o histograma da imagem
    for (int t = 0; t < num_threads; t++) {
        int inicio = t * linhas_por_thread;
        int fim = (t == num_threads - 1) ? altura_imagem : (t + 1) * linhas_por_thread; // A última thread processa o restante
        threads[t] = thread(correlacao_parcial, ref(imagem_entrada), ref(mascara), ref(imagem_saida), inicio, fim);
    }

    for (int t = 0; t < num_threads; t++) 
        threads[t].join();
    
    return imagem_saida;
}

Mat gerar_mascara_media(const int dimensao) {
    Mat mascara = Mat::zeros(dimensao, dimensao, CV_32F); 

    const float media = 1/(float)(dimensao * dimensao);

    for (int i = 0; i < dimensao; i++)
        for (int j = 0; j < dimensao; j++)
            mascara.at<float>(i, j) = media;

    return mascara; 
}